nvflare/app_common/widgets/client_api_configurator.py

```diff
@@ -0,0 +1,124 @@
+
+    @classmethod
+    def get_external_pipe_class(cls, pipe: Pipe) -> str:
+        pipe_class = type(pipe)
+        for pipe_type, info in cls.PIPE_EXTERNAL_MAPPINGS.items():
+            if issubclass(pipe_class, pipe_type):
+                return info[EXTERNAL_CLASS]
+        raise RuntimeError(f"Pipe of type ({pipe_class}) is not supported")
+
+    @classmethod
+    def get_external_pipe_args(cls, pipe: Pipe) -> dict:
+        pipe_class = type(pipe)
+        for pipe_type, info in cls.PIPE_EXTERNAL_MAPPINGS.items():
+            if issubclass(pipe_class, pipe_type):
+                args_mapping = info[EXTERNAL_ARGS_MAPPING]
+                return {arg: mapper(pipe) for arg, mapper in args_mapping.items()}
+        raise RuntimeError(f"Pipe of type ({pipe_class}) is not supported")
+
+    @classmethod
+    def write_config_to_file(cls, config_data: dict, fl_ctx: FLContext, filename: str = CLIENT_API_CONFIG):
+        """Writes client api config file into app_directory/config folder."""
+        engine = fl_ctx.get_engine()
+        workspace = engine.get_workspace()
+        app_directory = workspace.get_app_dir(fl_ctx.get_job_id())
+        config_file_path = os.path.join(app_directory, workspace.config_folder, filename)
+        if os.path.exists(config_file_path):
+            client_config = from_file(config_file=config_file_path)
+        else:
+            client_config = ClientConfig()
+        configuration = client_config.config
+        configuration[ConfigKey.SITE_NAME] = fl_ctx.get_identity_name()
+        configuration[ConfigKey.JOB_ID] = fl_ctx.get_job_id()
+        configuration.update(config_data)
+        client_config.to_json(config_file_path)
+
+    def __init__(self, component_ids: List[str]):
+        super().__init__()
+        check_object_type("component_ids", component_ids, list)
+
+        # the components that needs to provide information to the client api
+        self._component_ids = component_ids
+
+    def handle_event(self, event_type: str, fl_ctx: FLContext):
+        if event_type == EventType.ABOUT_TO_START_RUN:
+            components_data = self._export_all_components_for_client_api(fl_ctx)
+            ClientAPIConfigurator.write_config_to_file(config_data=components_data, fl_ctx=fl_ctx)
+
+    def _export_all_components_for_client_api(self, fl_ctx: FLContext) -> dict:
+        """Exports all components for client api."""
+        engine = fl_ctx.get_engine()
+        components = engine.get_all_components()
+        components_data = {}
+        for component_id in self._component_ids:
+            component_instance: ClientAPIExportable = components.get(component_id)
+            if component_instance is None:
+                raise RuntimeError(f"missing component: {component_id}")
+            component_attributes = component_instance.export_for_client_api()
+            for k, v in component_attributes.items():
+                if k in components_data:
+                    raise RuntimeError(
```

Comment:
I don't think we can require all components to have globally unique attributes.
The config structure should have separate sections for different components. It should be okay for different components to use the same attr names.

Reply:
thanks, fixed as suggested